streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -308,12 +305,41 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
         supervisor.start()
         supervisor.awaitTermination()
       }
+
       // Run the dummy Spark job to ensure that all slaves have registered.
       // This avoids all the receivers to be scheduled on the same node.
       if (!ssc.sparkContext.isLocal) {
         ssc.sparkContext.makeRDD(1 to 50, 50).map(x => (x, 1)).reduceByKey(_ + _, 20).collect()
       }
 
+      // Right now, we only honor preferences if all receivers have them
+      val hasLocationPreferences = receivers.map(_.preferredLocation.isDefined).reduce(_ && _)
+
+      // If no location preferences are specified, set host location for each receiver
+      // so as to distribute them evenly over executors in a round-robin fashion
+      var roundRobin = false;
+      if (!hasLocationPreferences && !ssc.sparkContext.isLocal) {
+        val executors = getExecutors(ssc)
+        if (!executors.isEmpty) {
+          var i = 0;
+          for (i <- 0 to (receivers.length - 1)) {
+            receivers(i).host = Some(executors(i % executors.length))
+          }
+          roundRobin = true
+        }
+      }
+
+      val tempRDD =
+        if (hasLocationPreferences) {
```

Comment:
Because the non-polling version of Flume receiver NEEDS a location (so that Flume can push events to that machine), whereas no other sources that we need requires it. We dont need to encourage anything, just good to support it if any one wants to do it. In fact, between the two Flume approaches, the push-based approach referred here is not a good approach, but still is present because the other pull-based polling approach is slightly complicated to set up and may not be feasible for some settings. So we HAVE to support both. For Kafka, we dont need to set preferred location.